gossip/tx_scrambler_test.go

```diff
@@ -0,0 +1,215 @@
+	}
+}
+
+func TestTxScrambler_UnifyEntries_ReportsDuplicateAddresses(t *testing.T) {
+	entries := createScramblerTestInputWithRepeatedAddresses()
+	_, _, hasDuplicateAddresses := unifyEntries(entries)
+	if !hasDuplicateAddresses {
+		t.Error("entries have duplicate addresses")
+	}
+
+	entries = createScramblerTestInputWithRepeatedHashes()
+	_, _, hasDuplicateAddresses = unifyEntries(entries)
+	if hasDuplicateAddresses {
+		t.Error("entries does not have duplicate addresses")
+	}
+}
+
+func TestTxScrambler_ScrambleTransactions_ScrambleIsDeterministic(t *testing.T) {
+	res1 := createScramblerTestInputWithRepeatedAddresses()
+	res2 := deepCopyEntries(res1)
+	// shuffle one array
+	shuffleEntries(res2)
+
+	salt := [32]byte{1}
+	scrambleTransactions(res1, salt)
+	scrambleTransactions(res2, salt)
+	if !reflect.DeepEqual(res1, res2) {
+		t.Error("scramble is not deterministic")
+	}
+}
+
+func TestTxScrambler_SortTransactionsByNonce_SortsSameSenderByNonce(t *testing.T) {
+	entries := createScramblerTestInputWithRepeatedAddresses()
+	entries = sortTransactionsByNonce(entries)
+	for i := 0; i < len(entries); i++ {
+		for j := i + 1; j < len(entries); j++ {
+			if entries[i].sender == entries[j].sender {
+				if entries[i].nonce > entries[j].nonce {
+					t.Errorf("incorrect nonce order %d must be before %d", entries[j].nonce, entries[i].nonce)
+				}
+			}
+		}
+	}
+}
+
+func TestTxScrambler_GetExecutionOrder_SortIsDeterministic(t *testing.T) {
+	res1 := createRandomScramblerTestInput()
+	res2 := deepCopyEntries(res1)
+	// shuffle one array
+	shuffleEntries(res2)
+
+	res1 = getExecutionOrder(res1)
+	res2 = getExecutionOrder(res2)
+	if !reflect.DeepEqual(res1, res2) {
+		t.Fatal("slices have different order - algorithm is not deterministic")
+	}
+}
+
+func TestTxScrambler_GetExecutionOrder_SortRemovesDuplicateHashes(t *testing.T) {
+	const size = 10
+	entries := createScramblerTestInputWithRepeatedHashes()
+	shuffleEntries(entries)
+
+	entries = getExecutionOrder(entries)
+
+	if got, want := len(entries), size/2; got != want {
+		t.Errorf("duplicate entries does not seem to be removed, got size: %v, want size: %v", got, want)
+	}
+}
+
+func TestTxScrambler_GetExecutionOrder_SortsSameSenderByNonce(t *testing.T) {
+	entries := createScramblerTestInputWithRepeatedAddresses()
+	entries = getExecutionOrder(entries)
+
+	for i := 0; i < len(entries); i++ {
+		for j := i + 1; j < len(entries); j++ {
+			if entries[i].sender == entries[j].sender {
+				if entries[i].nonce > entries[j].nonce {
+					t.Errorf("incorrect nonce order %d must be before %d", entries[j].nonce, entries[i].nonce)
+				}
+			}
+		}
+	}
+}
+
+func BenchmarkTxScrambler(b *testing.B) {
+	const (
+		numLoops   = 4
+		multiplier = 10
+	)
+	size := int64(10)
+
+	for _ = range numLoops {
+		b.Run(fmt.Sprintf("builtIn_%d", size), func(b *testing.B) {
+			for i := 1; i <= b.N; i++ {
+				b.StopTimer()
+				entries := createRandomScramblerTestInput()
+				b.StartTimer()
+				getExecutionOrder(entries)
```

Comment:
The `size` is nowhere used in this benchmark ... it should be odd that the benchmark probably reports the same time for any size of input.